### PR TITLE
Add the md5sum field to parquet filenames

### PR DESCRIPTION
**Pull request by RanMaoyi**

### CODE REVIEW

**Comment by JaylinYu** [comment]
Need to extract md5 info from filename in send_msg_file

File: include/nng/supplemental/nanolib/md5.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	a += b; \
+}
+#define II(a,b,c,d,x,s,ac) \
+{ \
+	a += I(b,c,d) + x + ac; \
+	a = ROTATE_LEFT(a,s); \
+	a += b; \
+}                                            
+
+int Compute_string_md5(unsigned char *dest_str, unsigned int dest_len, char *md5_str);
```
Comment (JaylinYu) [inline include/nng/supplemental/nanolib/md5.h:59]: unaligned func naming

**Review by JaylinYu — APPROVED** [review]

**Comment by JaylinYu** [comment]
Whoops md5.h included unistd.h

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 2]
```diff
@@ -490,6 +484,46 @@ parquet_write(
...
+
+	strcat(md5_file_name, "_");
+	strcat(md5_file_name, md5_buffer);
+	strcat(md5_file_name, filename + strlen(conf->dir) + strlen(conf->file_name_prefix) + 1);
+
+	ret = rename(filename, md5_file_name);
+	if (ret != 0) {
+		log_error("Failed to rename file %s to %s errno: %d", filename, md5_file_name, errno);
+		free(filename);
+		free(md5_file_name);
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:513]: Need to pthread_mutex_unlock(&parquet_queue_mutex);

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 3]
```diff
@@ -490,6 +484,46 @@ parquet_write(
...
+	if (ret != 0) {
+		log_error("Failed to calculate md5sum");
+		free(filename);
+		return -1;
+	}
+
+	char *md5_file_name = (char *)malloc(strlen(filename) + strlen("_") + strlen(md5_buffer) + 2);
+	if (md5_file_name == NULL) {
+		log_error("Failed to allocate memory for file name.");
+		free(filename);
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:498]: Need to pthread_mutex_unlock(&parquet_queue_mutex);

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 4]
```diff
@@ -490,6 +484,46 @@ parquet_write(
 			goto again;
 	}
 
+	char md5_buffer[33];
+	int ret = Compute_file_md5(filename, md5_buffer);
+	if (ret != 0) {
+		log_error("Failed to calculate md5sum");
+		free(filename);
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:491]: need to pthread_mutex_unlock(&parquet_queue_mutex);

**Review by wanghaEMQ — CHANGES_REQUESTED** [review]

File: src/supplemental/nanolib/parquet/parquet.cc [hunk 5]
```diff
@@ -490,6 +484,46 @@ parquet_write(
...
+		free(md5_file_name);
+		return -1;
+	}
+
+	free(filename);
+
+	ENQUEUE(parquet_file_queue, md5_file_name);
+
+	if (QUEUE_SIZE(parquet_file_queue) > conf->file_count) {
+		remove_old_file();
```
Comment (wanghaEMQ) [inline src/supplemental/nanolib/parquet/parquet.cc:522]: One more question. If any of the processes of renaming or calculating md5 failed the file counter will exceed the file_count in configuration.

Comment (RanMaoyi) [reply src/supplemental/nanolib/parquet/parquet.cc:522]: I'm not sure what you mean, if the renaming and calculating md5 fails there will be no enqueue action, the file_count will not grow, and there will be no situation where the file_count exceeds the configured value.

File: src/supplemental/nanolib/parquet/parquet.cc (hunk 2 above)
Comment (RanMaoyi) [reply src/supplemental/nanolib/parquet/parquet.cc:517]: thx, done

File: src/supplemental/nanolib/parquet/parquet.cc (hunk 4 above)
Comment (RanMaoyi) [reply src/supplemental/nanolib/parquet/parquet.cc:491]: thx, done

File: include/nng/supplemental/nanolib/md5.h (hunk 1 above)
Comment (RanMaoyi) [reply include/nng/supplemental/nanolib/md5.h]: Rename to ComputeStringMD5()

**Comment by RanMaoyi** [comment]
> Whoops md5.h included unistd.h

unistd.h needed by read() etc...

File: src/supplemental/nanolib/parquet/parquet.cc (hunk 5 above)
Comment (wanghaEMQ) [reply src/supplemental/nanolib/parquet/parquet.cc:528]: Not the length of the queue. I mean the number of files written to disk

Comment (RanMaoyi) [reply src/supplemental/nanolib/parquet/parquet.cc:543]: Got that, already remove file when calculating md5 or renaming failed

**Review by JaylinYu — APPROVED** [review]

**Pull request merged**